src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,694 @@
+        );
+
+        if keymap.is_null() {
+            panic!("Received invalid keymap from compositor.");
+        }
+
+        self.post_init(keymap);
+    }
+
+    pub(crate) unsafe fn init_with_rmlvo(
+        &mut self,
+        names: ffi::xkb_rule_names,
+    ) -> Result<(), Error> {
+        let keymap = (XKBH.xkb_keymap_new_from_names)(
+            self.xkb_context,
+            &names,
+            ffi::xkb_keymap_compile_flags::XKB_KEYMAP_COMPILE_NO_FLAGS,
+        );
+
+        if keymap.is_null() {
+            return Err(Error::BadNames);
+        }
+
+        self.post_init(keymap);
+
+        Ok(())
+    }
+}
+
+impl KbState {
+    pub(crate) unsafe fn key_repeats(&mut self, keycode: ffi::xkb_keycode_t) -> bool {
+        (XKBH.xkb_keymap_key_repeats)(self.xkb_keymap, keycode + 8) == 1
+    }
+
+    #[inline]
+    pub(crate) fn ready(&self) -> bool {
+        !self.xkb_state.is_null()
+    }
+
+    #[inline]
+    pub(crate) fn locked(&self) -> bool {
+        self.locked
+    }
+
+    #[inline]
+    pub(crate) fn mods_state(&self) -> ModifiersState {
+        self.mods_state
+    }
+}
+
+impl Drop for KbState {
+    fn drop(&mut self) {
+        unsafe {
+            (XKBCH.xkb_compose_state_unref)(self.xkb_compose_state);
+            (XKBCH.xkb_compose_table_unref)(self.xkb_compose_table);
+            (XKBH.xkb_state_unref)(self.xkb_state);
+            (XKBH.xkb_keymap_unref)(self.xkb_keymap);
+            (XKBH.xkb_context_unref)(self.xkb_context);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    /// libxkbcommon is not available
+    XKBNotFound,
+    /// Provided RMLVO specified a keymap that would not be loaded
+    BadNames,
+}
+
+impl KbState {
+    pub fn process_key_event(&mut self, keycode: u32, state: ElementState) -> KeyEventResults<'_> {
+        KeyEventResults::new(self, keycode, state == ElementState::Pressed)
+    }
+
+    pub fn process_key_repeat_event(&mut self, keycode: u32) -> KeyEventResults<'_> {
+        KeyEventResults::new(self, keycode, false)
+    }
+
+    fn keysym_to_utf8_raw(&mut self, keysym: u32) -> Option<&'static str> {
+        self.scratch_buffer.clear();
+        self.scratch_buffer.reserve(8);
+        loop {
+            unsafe { self.scratch_buffer.set_len(8) };
+            let bytes_written = unsafe {
+                (XKBH.xkb_keysym_to_utf8)(
+                    keysym,
+                    self.scratch_buffer.as_mut_ptr().cast(),
+                    self.scratch_buffer.capacity(),
+                )
+            };
+            if bytes_written == 0 {
+                return None;
+            } else if bytes_written == -1 {
+                self.scratch_buffer.reserve(8);
+            } else {
+                unsafe {
+                    self.scratch_buffer
+                        .set_len(bytes_written.try_into().unwrap())
+                };
+                break;
+            }
+        }
```

Comment:
> I'm assuming that you originally intended to call `xkb_keysym_to_utf8` one more time after `bytes_written == -1`. However I'm willing to bet that there's no measurable difference between allocating 8 vs allocating 16 bytes for the first attempt (especially because you are using this neat `scratch_buffer`). So maybe just allocate big, but once, and log an error if the buffer is not large enough.

8 bytes was honestly chosen pretty arbitrarily, as far as I can recall. I'd like to retain the ability to handle a UTF-8 string of an arbitrary length, as that's more "future proof". I think there's supposed to be a trick that lets you figure out how many bytes you need in one go, but I can't recall why I didn't use that in the first place. I need to revisit this function anyway, as I think I may be passing along a null terminator by accident here.

> Also the `set_len` before the `xkb_keysym_to_utf8` is not needed, in fact I think it should be removed considering you are setting the length afterwards.

Good catch! It's completely superfluous.

Reply:
After reading up on this again, the decisions I made seem to make sense now as well.

The docs say that the buffer passed in must be at least 7 bytes long, and the function returns `-1` when the buffer is too small, hence why I loop in order to eventually get a big enough buffer.

Reply:
Oh I think I completely missed the `loop` around it. It does seem to make sense.